synapse/storage/databases/main/roommember.py

```diff
@@ -290,6 +305,24 @@ async def get_invite_for_local_user_in_room(
                 return invite
         return None
 
+    async def get_knock_for_local_user_in_room(
+        self, user_id: str, room_id: str
+    ) -> Optional[RoomsForUser]:
+        """Gets the knock for the given *local* user and room.
+
+        Args:
+            user_id: The user ID to find the knock of.
+            room_id: The room to user knocked on.
+
+        Returns:
+            Either a RoomsForUser or None if no knock was found.
+        """
+        knocks = await self.get_knocked_rooms_for_local_user(user_id)
+        for knock in knocks:
+            if knock.room_id == room_id:
+                return knock
```

Comment:
I see that this just matches what is doing for invites, but this seems inefficient if we know the room ID that we first fetch the full list of rooms?! I'm guessing this is so that we hit the cache instead of the database?

Reply:
I've attempted a more efficient approach in ad929c3bf, and have applied the same approach to the invite code that was originally cribbed off of. Tests failed, I fixed things up, and they passed again, which is a good sign.

Let me know what you think, or if the invite change should be moved to another PR.